src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -77,68 +77,93 @@ private async Task<bool> IsEnabledAsync<TContext>(string feature, TContext appCo
             if (featureDefinition != null)
             {
                 //
-                // Check if feature is always on
-                // If it is, result is true, goto: cache
+                // If the requirement type is Any, we end on a true. Requirement type All will end on a false
+                bool targetEvaluation = featureDefinition.RequirementType == RequirementType.Any;
 
-                if (featureDefinition.EnabledFor.Any(featureFilter => string.Equals(featureFilter.Name, "AlwaysOn", StringComparison.OrdinalIgnoreCase)))
-                {
-                    enabled = true;
-                }
-                else
+                //
+                // For all enabling filters listed in the feature's state, evaluate them according to requirement type
+                foreach (FeatureFilterConfiguration featureFilterConfiguration in featureDefinition.EnabledFor)
                 {
                     //
-                    // For all enabling filters listed in the feature's state calculate if they return true
-                    // If any executed filters return true, return true
+                    // If the filter is AlwaysOn and we're targeting true, return true
+                    if (string.Equals(featureFilterConfiguration.Name, "AlwaysOn", StringComparison.OrdinalIgnoreCase) && targetEvaluation)
+                    {
+                        enabled = true;
 
-                    foreach (FeatureFilterConfiguration featureFilterConfiguration in featureDefinition.EnabledFor)
+                        break;
+                    }
+
+                    IFeatureFilterMetadata filter = GetFeatureFilterMetadata(featureFilterConfiguration.Name);
+
+                    if (filter == null)
                     {
-                        IFeatureFilterMetadata filter = GetFeatureFilterMetadata(featureFilterConfiguration.Name);
+                        string errorMessage = $"The feature filter '{featureFilterConfiguration.Name}' specified for feature '{feature}' was not found.";
 
-                        if (filter == null)
+                        if (!_options.IgnoreMissingFeatureFilters)
                         {
-                            string errorMessage = $"The feature filter '{featureFilterConfiguration.Name}' specified for feature '{feature}' was not found.";
-
-                            if (!_options.IgnoreMissingFeatureFilters)
+                            throw new FeatureManagementException(FeatureManagementError.MissingFeatureFilter, errorMessage);
+                        }
+                        else
+                        {
+                            if (featureDefinition.RequirementType == RequirementType.All)
                             {
+                                errorMessage = $"The 'IgnoreMissingFeatureFilters' flag cannot override a missing filter on a feature with requirement type 'All'. {errorMessage}";
```

Comment:
I think we should block this combination even if filters are not missing. It produces a more deterministic restriction which is not dependent on users' feature filters, which may change in future.